### ISSUE TITLE
Auto-derivation of Show instances

### ISSUE THREAD

**Issue by lexi-lambda** [lexi-lambda/hackett#8]
This can be a macro specific to `Show` for now, and eventually it can be generalized to support more kinds of auto-derivation.

**Comment by johanatan** [comment]
This seems like a good place for me to dive into the codebase. Would you mind sharing a few pointers into the code on where/what needs to happen for this one?

**Comment by lexi-lambda** [comment]
Sure. You probably want to start by writing a macro that works something like this:

```racket
(data (X a) (A Integer) (B a String))
(derive-instance/Show X)
```

…and produces an instance that looks like this:

```racket
(instance (forall [a] (Show a) => (Show (X a)))
  [show (lambda* [[(A a)] {"(A " ++ (show a) ++ ")"}]
                 [[(B a b)] {"(B " ++ (show a) ++ " " ++ (show b) ++ ")"}])])
```

The most relevant module in Hackett is `hackett/private/adt`. Unfortunately, some of this code is a bit hairy, and it might need a little bit of restructuring. Currently, datatype names are bound to type transformers directly, but this likely isn’t the right thing to do… more probably, they ought to be bound to some structure that keeps track of information about the datatype for macros to consume.

Are you at all familiar with the Racket macro system, by any chance? If not, it might be worth getting familiar with some of the fancier parts before diving into the guts of Hackett’s implementation.

**Comment by lexi-lambda** [comment]
Closed in 5866745003cd1814a1316057ccc0399b85bc5aea and 55f32e8b6228207e9af923a791318e5d4c056630.
